### PR TITLE
Refs #29253 -- Fixed method_decorator() crash if decorator sets a new attribute.

### PR DESCRIPTION
**Pull request by cjerdonek**
This fixes a regression stated in PR #9819 in this comment:
https://github.com/django/django/pull/9819#issuecomment-399437486

### CODE REVIEW

File: django/utils/decorators.py [hunk 1]
```diff
@@ -36,8 +36,13 @@ def _multi_decorate(decorators, method):
...
     def _wrapper(self, *args, **kwargs):
         # bound_method has the signature that 'decorator' expects i.e. no
-        # 'self' argument.
-        bound_method = method.__get__(self, type(self))
+        # 'self' argument, but it is a closure over self so it can call
+        # 'func' correctly.  Also, we need to wrap method.__get__() in
+        # a function because one can't set new attributes on bound method
+        # objects -- only functions.
+        def bound_method(*args2, **kwargs2):
+            return method.__get__(self, type(self))(*args2, **kwargs2)
```
Comment (charettes) [inline django/utils/decorators.py]: Would `functools.partial` work here?

```python
from functools import partial
bound_method = partial(method.__get__(self, type(self)))
```

File: tests/decorators/tests.py [hunk 2]
```diff
@@ -271,6 +271,36 @@ def method(self):
...
+                pass
+
+        obj = MyClass()
+        obj.method()  # This should not error out.
+
+    def test_non_atomic_requests(self):
+        """
+        Test transaction.non_atomic_requests().
+        """
+        import django.db.transaction as transaction
```
Comment (charettes) [inline tests/decorators/tests.py]: Move to the top.

File: tests/decorators/tests.py [hunk 3]
```diff
@@ -271,6 +271,36 @@ def method(self):
...
+
+        class MyClass:
+            @method_decorator(decorate)
+            def method(self):
+                pass
+
+        obj = MyClass()
+        obj.method()  # This should not error out.
+
+    def test_non_atomic_requests(self):
```
Comment (charettes) [inline tests/decorators/tests.py]: I don't think both tests are required.

**Comment by cjerdonek** [comment]
I addressed @charettes's comments.

**Pull request merged**